File: json/json-annotation-processor/src/test/java/io/koraframework/json/annotation/processor/DelegatingValueTest.java
package io.koraframework.json.annotation.processor;

import org.junit.jupiter.api.Test;
import io.koraframework.json.common.JsonReader;
import io.koraframework.json.common.JsonWriter;
import tools.jackson.core.JsonGenerator;
import tools.jackson.core.JsonParser;

import java.nio.charset.StandardCharsets;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class DelegatingValueTest extends AbstractJsonAnnotationProcessorTest {
    JsonReader<Long> longReader = JsonParser::getLongValue;
    JsonWriter<Long> longWriter = JsonGenerator::writeNumber;
    JsonReader<String> stringReader = JsonParser::getValueAsString;
    JsonWriter<String> stringWriter = JsonGenerator::writeString;

    private Object newObject(String name, Class<?> argType, Object arg) throws Exception {
        return compileResult.loadClass(name).getDeclaredConstructor(argType).newInstance(arg);
    }

    @Test
    public void testDelegatingInstanceWriterAndFactoryReader() throws Exception {
        compile("""
            public record UserId(long id) {
              @JsonReader public static UserId of(long v) { return new UserId(v); }
              @JsonWriter public long id() { return id; }
            }
            """);
        compileResult.assertSuccess();
        var mapper = mapper("UserId", List.of(longReader), List.of(longWriter));
        mapper.verify(newObject("UserId", long.class, 42L), "42");
    }

    @Test
    public void testDelegatingStaticWriter() throws Exception {
        compile("""
            public record UserId(long id) {
              @JsonReader public static UserId of(long v) { return new UserId(v); }
              @JsonWriter public static long toJson(UserId u) { return u.id(); }
            }
            """);
        compileResult.assertSuccess();
        var mapper = mapper("UserId", List.of(longReader), List.of(longWriter));
        mapper.verify(newObject("UserId", long.class, 7L), "7");
    }

    @Test
    public void testDelegatingStringValue() throws Exception {
        compile("""
            public record Sku(String code) {
              @JsonReader public static Sku parse(String v) { return new Sku(v); }
              @JsonWriter public String code() { return code; }
            }
            """);
        compileResult.assertSuccess();
        var mapper = mapper("Sku", List.of(stringReader), List.of(stringWriter));
        mapper.verify(newObject("Sku", String.class, "ABC"), "\"ABC\"");
    }

    @Test
    public void testDelegatingWriterNullHandling() throws Exception {
        compile("""
            public record Sku(String code) {
              @JsonReader public static Sku parse(String v) { return new Sku(v); }
              @JsonWriter public String code() { return code; }
            }
            """);
        compileResult.assertSuccess();
        var w = writer("Sku", stringWriter);
        assertThat(w.toByteArray(null)).asString(StandardCharsets.UTF_8).isEqualTo("null");
    }
}

File: json/json-symbol-processor/src/test/kotlin/io/koraframework/json/ksp/DelegatingValueTest.kt
package io.koraframework.json.ksp

import io.koraframework.json.common.JsonReader
import io.koraframework.json.common.JsonWriter
import org.junit.jupiter.api.Test
import tools.jackson.core.JsonGenerator
import tools.jackson.core.JsonParser

class DelegatingValueTest : AbstractJsonSymbolProcessorTest() {
    private val stringReader = JsonReader<String> { p: JsonParser -> p.valueAsString }
    private val stringWriter = JsonWriter<String> { g: JsonGenerator, v: String? -> g.writeString(v) }
    private val longReader = JsonReader<Long> { p: JsonParser -> p.longValue }
    private val longWriter = JsonWriter<Long> { g: JsonGenerator, v: Long? -> g.writeNumber(v!!) }

    private fun newObject(name: String, argType: Class<*>, arg: Any?): Any =
        compileResult.assertSuccess().classLoader.loadClass("${testPackage()}.$name").getDeclaredConstructor(argType).newInstance(arg)

    @Test
    fun testDelegatingInstanceWriterAndFactoryReader() {
        compile(
            """
            class UserId(val id: Long) {
              @JsonWriter fun toJson(): Long = id
              companion object { @JsonReader fun of(v: Long): UserId = UserId(v) }
              override fun equals(other: Any?) = other is UserId && other.id == id
              override fun hashCode() = id.hashCode()
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val mapper = mapper("UserId", listOf(longReader), listOf(longWriter))
        mapper.assert(newObject("UserId", Long::class.javaPrimitiveType!!, 42L), "42")
    }

    @Test
    fun testDelegatingStaticWriter() {
        compile(
            """
            class UserId(val id: Long) {
              companion object {
                @JsonReader fun of(v: Long): UserId = UserId(v)
                @JsonWriter fun toJson(u: UserId): Long = u.id
              }
              override fun equals(other: Any?) = other is UserId && other.id == id
              override fun hashCode() = id.hashCode()
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val mapper = mapper("UserId", listOf(longReader), listOf(longWriter))
        mapper.assert(newObject("UserId", Long::class.javaPrimitiveType!!, 7L), "7")
    }

    @Test
    fun testDelegatingStringValue() {
        compile(
            """
            class Sku(val code: String) {
              @JsonWriter fun toJson(): String = code
              companion object { @JsonReader fun parse(v: String): Sku = Sku(v) }
              override fun equals(other: Any?) = other is Sku && other.code == code
              override fun hashCode() = code.hashCode()
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val mapper = mapper("Sku", listOf(stringReader), listOf(stringWriter))
        mapper.assert(newObject("Sku", String::class.java, "ABC"), "\"ABC\"")
    }
}
